sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/EventTimeWatermarkWithWatermarkDefInSelectSuite.scala

```diff
@@ -0,0 +1,267 @@
+  with BeforeAndAfter
+  with Logging {
+
+  import testImplicits._
+
+  after {
+    sqlContext.streams.active.foreach(_.stop())
+  }
+
+  test("event time and watermark metrics with watermark in select DML - case 1") {
+    // All event time metrics where watermarking is set
+    val inputData = MemoryStream[Int]
+    val df = inputData.toDF()
+      .withColumn("eventTime", timestamp_seconds(col("value")))
+    df.createOrReplaceTempView("stream_src")
+    val aggWithWatermark = spark.sql(
+      """
+        |SELECT
+        |    CAST(window.start AS LONG), CAST(count(*) AS LONG) AS count
+        |FROM
+        |    stream_src WATERMARK eventTime DELAY OF INTERVAL 10 seconds
+        |GROUP BY window(eventTime, '5 seconds')
+        |""".stripMargin)
+
+    testWindowedAggregation(inputData, aggWithWatermark)
+  }
+
+  test("event time and watermark metrics with watermark in select DML - case 2") {
+    // All event time metrics where watermarking is set
+    val inputData = MemoryStream[Int]
+    val df = inputData.toDF()
+    df.createOrReplaceTempView("stream_src")
+    val aggWithWatermark = spark.sql(
+      """
+        |SELECT
+        |    CAST(window.start AS LONG), CAST(count(*) AS LONG) AS count
+        |FROM
+        |    stream_src
+        |    WATERMARK timestamp_seconds(value) AS eventTime DELAY OF INTERVAL 10 seconds
+        |GROUP BY window(eventTime, '5 seconds')
+        |""".stripMargin)
+
+    testWindowedAggregation(inputData, aggWithWatermark)
+  }
+
+  private def testWindowedAggregation(
+      inputData: MemoryStream[Int],
+      dataFrame: DataFrame): Unit = {
+    testStream(dataFrame)(
+      AddData(inputData, 15),
+      CheckAnswer(),
+      assertEventStats(min = 15, max = 15, avg = 15, wtrmark = 0),
+      AddData(inputData, 10, 12, 14),
+      CheckAnswer(),
+      assertEventStats(min = 10, max = 14, avg = 12, wtrmark = 5),
+      AddData(inputData, 25),
+      CheckAnswer((10, 3)),
+      assertEventStats(min = 25, max = 25, avg = 25, wtrmark = 5)
+    )
+  }
+
+  test("stream-stream join with watermark in select DML - case 1") {
+    val leftInput = MemoryStream[(Int, Int)]
+    val rightInput = MemoryStream[(Int, Int)]
+
+    val df1 = leftInput.toDF().toDF("leftKey", "time")
+      .select($"leftKey", timestamp_seconds($"time") as "leftTime",
+        ($"leftKey" * 2) as "leftValue")
+    val df2 = rightInput.toDF().toDF("rightKey", "time")
+      .select($"rightKey", timestamp_seconds($"time") as "rightTime",
+        ($"rightKey" * 3) as "rightValue")
+
+    df1.createOrReplaceTempView("stream_left")
+    df2.createOrReplaceTempView("stream_right")
+
+    val joined = spark.sql(
+      """
+        |SELECT
+        |    leftKey, rightKey, CAST(leftTime AS INTEGER), CAST(rightTime AS INTEGER)
+        |FROM
+        |    stream_left WATERMARK leftTime DELAY OF INTERVAL 0 second
+        |FULL OUTER JOIN
+        |    stream_right WATERMARK rightTime DELAY OF INTERVAL 0 second
```

Comment:
do we require the watermark to be the same between two join sides? what's the corresponding dataframe query?

Reply:
I have the impression that, with DataFrame API, we can set watermark for all streaming operators, not only the streaming scans. Is this true?

Reply:
You can set watermark "before" the first stateful operator has appeared. But conceptually, it is advised to define watermark closer to stream scan. The concept of watermark is to mark that the engine has seen (processed) all the data before the timestamp ts, which is closely related to the input data from stream scan.